mkosi/distributions/debian.py

```diff
@@ -190,6 +191,68 @@ def install(cls, context: Context) -> None:
+        # match what we are building, and set up a diversion so that dpkg doesn't overwrite it on package updates.
+        for candidate in ["etc/os-release", "usr/lib/os-release", "usr/lib/initrd-release"]:
+            osrelease = context.root / candidate
+            # at this point we know we will either change or add to the file
+            newosrelease = osrelease.with_suffix(".new")
+
+            if not osrelease.is_file():
+                continue
+
+            if osrelease.is_symlink() and candidate != "etc/os-release":
+                continue
+
+            image_id_written = image_version_written = False
+            with osrelease.open("r") as old, newosrelease.open("w") as new:
+                # fix existing values
+                for line in old.readlines():
+                    if context.config.image_id and line.startswith("IMAGE_ID="):
+                        new.write(f'IMAGE_ID="{context.config.image_id}"\n')
+                        image_id_written = True
+                    elif context.config.image_version and line.startswith("IMAGE_VERSION="):
+                        new.write(f'IMAGE_VERSION="{context.config.image_version}"\n')
+                        image_version_written = True
+                    elif (
+                        context.config.release in ("unstable", "sid") and
+                        line.startswith("VERSION_CODENAME=")
+                    ):
+                        # Workaround for Debian brokeness: unstable sets VERSION_CODENAME=<testing codename>
+                        new.write('VERSION_CODENAME=sid\n')
+                    else:
+                        new.write(line)
+
+                # append if they were missing
+                if context.config.image_id and not image_id_written:
+                    new.write(f'IMAGE_ID="{context.config.image_id}"\n')
+                if context.config.image_version and not image_version_written:
+                    new.write(f'IMAGE_VERSION="{context.config.image_version}"\n')
+
+            # On dpkg distributions we cannot simply overwrite the file as it is owned by a package.
+            # We need to set up a diversion first, so that it is not overwritten by package updates.
+            # We do this for the file in /etc/ as that will be overwritten on package updates and has
+            # precedence over the file in /usr/lib/, and ignore the latter and assume that if an usr-only
+            # image is built then the package manager will not run on it.
+            if candidate == "etc/os-release":
+                run([
+                    "dpkg-divert",
+                    "--quiet",
+                    "--root",
+                    context.root,
+                    "--local",
+                    "--add",
+                    "--rename",
+                    "--divert",
+                    f"/{candidate}.mkosi",
+                    f"/{candidate}"
+                ], stderr=subprocess.DEVNULL)
```

Comment:
This needs to run within a sandbox like all the other commands. Please specify a sandbox with `context.root` mounted to `/buildroot` and then have `dpkg-divert` operate on that.

Reply:
Please also open a bug for /etc/os-release being overridden by base-files updates and then link it here (I don't care if it gets immediately closed, it'll serve as documentation that all we're doing here is working around's debian's brokenness.